File: Top_100_Liked_Questions/621._Task_Scheduler/solution.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        # perform the most number of remaining task among those that are not in cooldown
        # we can also check the earilest possible itvl for any task to skip looping through idle
        
        # array of 26 characters with [count, earliest_interval]
        arr = [[0,0] for _ in range(26)]
        num_remaining_task = len(tasks)
        
        for task in tasks:
            arr[ord(task) - ord('A')][0] += 1
        
        itvl = 0
        while num_remaining_task:
            maxCnt = 0
            earilest_itvl = itvl + n
            maxCntTask = None
            
            for i in range(26):
                cnt, er_itvl = arr[i]
                if er_itvl <= itvl and cnt > maxCnt:
                    maxCnt = cnt
                    maxCntTask = i
                if cnt:
                    earilest_itvl = min(earilest_itvl, er_itvl)    
                    
            if maxCntTask is not None:
                arr[maxCntTask][0] -= 1
                num_remaining_task -= 1
                arr[maxCntTask][1] = itvl + n + 1

            itvl = max(itvl+1, earilest_itvl)
        
        return itvl

File: Top_100_Liked_Questions/621._Task_Scheduler/solution.js
/**
 * @param {character[]} tasks
 * @param {number} n
 * @return {number}
 */
var leastInterval = function(tasks, n) {
    let arr = new Array(26);
    for (let i=0; i<26; i++) {
        arr[i] = new Array(2).fill(0);
    }
    
    a_ascii = 'A'.charCodeAt(0);
    for (let task of tasks) {
        arr[task.charCodeAt(0) - a_ascii][0] += 1;
    }
    
    let rem_task = tasks.length;
    let itvl = 0;
    
    while (rem_task) {
        // get non-cooldown task with max number of remaining
        let max_cnt = 0;
        let max_task = null;
        let earliest_itvl = itvl + n;
        for (let i=0; i<26; i++) {
            let [cnt, er_itvl] = arr[i];
            if ((er_itvl <= itvl) && (cnt > max_cnt)) {
                max_cnt = cnt;
                max_task = i;
            }
            if (cnt) {
                earliest_itvl = Math.min(earliest_itvl, er_itvl);
            }
        }
        
        if (max_task != null) {
            arr[max_task][0]--;
            arr[max_task][1] = itvl + n + 1;
            rem_task--;
        }
        
        itvl = Math.max(itvl + 1, earliest_itvl);
    }
    
    return itvl;
};